idea/src/org/jetbrains/kotlin/idea/inspections/collections/SimplifiableCallChainInspection.kt

```diff
@@ -79,10 +82,14 @@ class SimplifiableCallChainInspection : AbstractCallChainChecker() {
     }
 }
 
-private fun KotlinType?.isMapOrSubType(): Boolean {
-    return this != null && (isMap() || constructor.supertypes.any { it.isMap() })
-}
-
-private fun KotlinType.isMap(): Boolean {
-    return constructor.declarationDescriptor?.name?.asString()?.endsWith("Map") == true
-}
+private fun KotlinType?.isMap(builtIns: KotlinBuiltIns): Boolean {
```

Comment:
@t-kameyama Why don't you just use the implementation I've suggested in my [first comment](https://github.com/JetBrains/kotlin/pull/1753#discussion_r204414873) in this PR? Is there anything wrong with it?
```
private fun KotlinType?.isMap(builtIns: KotlinBuiltIns): Boolean {
    return this != null && constructor.declarationDescriptor?.name?.asString()?.endsWith("Map") == true && isSubtypeOf(map)
}
```
(note that this is copied from https://github.com/JetBrains/kotlin/pull/1753#discussion_r204414873)

Reply:
If `type` is` HashMap`, `type.isSubtype (map)` returns false, so I think that recursive checks is necessary.

Reply:
It also returns false if `type` is` MutableMap`.

Reply:
The reason for this behavior is that you're taking the _default type_ of `kotlin.Map` (`builtIns.map.defaultType`). Default type of that class is `Map<K, V>` where `K` and `V` -- type parameters of `Map`. This type only makes sense inside the `Map` interface, as that is the only place where these type parameters can be used. Using this type for subtype checks is incorrect. Taking the example of `MutableMap`: `MutableMap<K, V>` is not a subtype of `Map<K, V>` because although the class `MutableMap` is a subclass of `Map`, the type arguments `K` and `V` in the left type are not the subtypes of the type arguments `K` and `V` in the right type, because they're different `K` and `V`: the ones on the left are from `MutableMap` and the ones on the right are from `Map`. Of course, they aren't compatible in any way because these are type parameters declared in different declarations, and there's no explicit relation between them. You can see how using default type makes little sense in this type of code.

I suggest you use `isSubclassOf` instead, which checks that out of any types in the hierarchy of the given type, _some_ type has the given class as the type constructor.